### PR TITLE
Internal data model for recommender

### PR DESCRIPTION
**Pull request by KarolKraskiewicz**
to be used by both metrics client and histogram aggregator

@mwielgus @kgrygiel - please reveiw

### CODE REVIEW

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+limitations under the License.
+*/
+
+package model
+
+import (
+	"time"
+)
+
+// MeasuredResourceName represent the name of the resource monitored by recommender
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: Maybe "MeasuredResourceName" => "MetricName" ?

Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "represents"

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+type MeasuredResourceName string
+
+const (
+	// ResourceCPU represents CPU in milicores (1core = 1000milicores)
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "mili" => "milli"

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	ResourceCPU MeasuredResourceName = "cpu"
+	// ResourceMemory represents memory, in bytes. (500Gi = 500GiB = 500 * 1024 * 1024 * 1024)
+	ResourceMemory MeasuredResourceName = "memory"
+)
+
+// PodID contains information needed to identify a Pod within a cluster
+type PodID struct {
+	// Namespaces where the Pod is defined
+	Namespace string
+	// PodName is a name of the pod unique within a namespace
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "the name"

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+// ContainerID contains information needed to identify a Container within a cluster
+type ContainerID struct {
+	PodID
+	// ContainerName is the name of the container, unique within a pod.
+	ContainerName string
+}
+
+// ContainerMetricsSnapshot contains information about usage of certain container withing defined time window
+type ContainerMetricsSnapshot struct { // nazwa zostaje
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: You have some trailing personal note here.

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	// ID identifies a specific container those metrics are comming from
+	ID ContainerID
+	// Time when the snapshot was taken
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "Start time of the measurement interval"

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	SnapshotTime time.Time
+	// duration, which this snapshot represents
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "Duration of the measurement interval"

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	SnapshotWindow time.Duration
+	// Actual usage of resources, during the SnapshotWindow.
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "Average usage of the resources over the measurement interval"

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	Usage map[MeasuredResourceName]int
+}
+
+// BasicPodSpec contains basic information defining a pod and its containers
+type BasicPodSpec struct {
+	// ID identifies a pod within a cluster
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: Finish the sentences with "." consistently.

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	ID PodID
+	// Labels of the pod. It is used to match pods with certain VPA opjects.
+	PodLabels map[string]string
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: How about using labels.Set ("k8s.io/apimachinery/pkg/labels")?

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	// List of containers within this pod
+	Containers []BasicContainerSpec
+}
+
+// BasicContainerSpec contains basic information defining a container
+type BasicContainerSpec struct {
+	// ID identifies a container within a cluster
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "the container"

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+// ContainerMetricsSnapshot contains information about usage of certain container withing defined time window
+type ContainerMetricsSnapshot struct { // nazwa zostaje
+	// ID identifies a specific container those metrics are comming from
+	ID ContainerID
+	// Time when the snapshot was taken
+	SnapshotTime time.Time
+	// duration, which this snapshot represents
+	SnapshotWindow time.Duration
+	// Actual usage of resources, during the SnapshotWindow.
+	Usage map[MeasuredResourceName]int
```
Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: Consider defining "type ResourceAmount int".

File: vertical-pod-autoscaler/recommender/model/types.go (hunk 7 above)
Comment (KarolKraskiewicz) [reply vertical-pod-autoscaler/recommender/model/types.go]: Good point, however, I was not able to find this information. Metrics api contains just types definitions and does not provide aggregation method, and API server, which we use, is pretty un-documented. 
(As a reference, heapster was using avg for cpu and current consumption for memory, but it is different mechanism here)

File: vertical-pod-autoscaler/recommender/model/types.go (hunk 9 above)
Comment (KarolKraskiewicz) [reply vertical-pod-autoscaler/recommender/model/types.go]: We already get map[string]string from the Lister API (see [here](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L209)). Re-packaging this map would only add an additional dependency to an external package.

File: vertical-pod-autoscaler/recommender/model/types.go [hunk 12]
```diff
@@ -20,59 +20,62 @@ import (
 	"time"
 )
 
-// MeasuredResourceName represent the name of the resource monitored by recommender
-type MeasuredResourceName string
+// MetricName represents the name of the resource monitored by recommender.
+type MetricName string
+
+// ResourceAmount represents quantity of a certain resource within a containr.
```
Comment (zync-perf) [inline vertical-pod-autoscaler/recommender/model/types.go]: "container"

Comment (kgrygiel) [inline vertical-pod-autoscaler/recommender/model/types.go]: "container"

**Comment by KarolKraskiewicz** [comment]
Type-o fixed and commits squashed.

**Pull request merged**